client/database.go

```diff
@@ -17,107 +17,212 @@ package client
+func (db *Database) createCollectionsFromSchemas(ctx context.Context, dbName string, schemas map[string]*schema.Schema) error {
 	tx, err := db.driver.BeginTx(ctx, dbName)
 	if err != nil {
 		return err
 	}
 	defer func() { _ = tx.Rollback(ctx) }()
 
+	created := make(map[string]bool)
+
 	for _, v := range schemas {
 		sch, err := schema.Build(v)
 		if err != nil {
 			return err
 		}
-		log.Debug().Interface("schema", v).Str("collection", v.Name).Msg("migrateModel")
 		err = tx.CreateOrUpdateCollection(ctx, v.Name, sch)
 		if err != nil {
 			return err
 		}
+		created[v.Name] = true
+	}
+
+	if err := tx.Commit(ctx); err != nil {
+		return err
+	}
+
+	db.created = created
+
+	return nil
+}
+
+// GetCollection returns collection object corresponding to collection model T
+func GetCollection[T schema.Model](ctx context.Context, db *Database) (*Collection[T], error) {
+	var m T
+	name := schema.ModelName(m)
+	ddb := db.driver.UseDatabase(db.name)
+
+	if !db.created[name] {
+		schema, err := schema.FromCollectionModel(m)
+		if err != nil {
+			return nil, err
+		}
+		sch, err := schema.Build()
+		if err != nil {
+			return nil, err
+		}
+		if err := ddb.CreateOrUpdateCollection(ctx, name, sch); err != nil {
+			return nil, err
+		}
+		db.created[name] = true
+		db.schemas[name] = schema
 	}
 
-	return tx.Commit(ctx)
+	return &Collection[T]{name: name, driver: db.driver, db: db,
+		schema: db.schemas[name], crud: ddb}, nil
+}
+
+// Create new Database
+func (db *Database) Create(ctx context.Context) error {
+	return db.driver.CreateDatabase(ctx, db.name)
+}
+
+// Drop Database.
+// All the collections in the Database will be dropped
+func (db *Database) Drop(ctx context.Context) error {
+	return db.driver.DropDatabase(ctx, db.name)
+}
+
+// Tx executes given set of operations in a transaction
+func (db *Database) Tx(ctx context.Context, fn func(ctx context.Context, tx *Tx) error) error {
+	dtx, err := db.driver.BeginTx(ctx, db.name)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = dtx.Rollback(ctx) }()
+
+	tx := &Tx{db, dtx}
+
+	if err = fn(ctx, tx); err != nil {
+		return err
+	}
+
+	return dtx.Commit(ctx)
+}
+
+// openDatabase create Database and collections from the provided Database model
+func openDatabase[T schema.Model](ctx context.Context, d driver.Driver) (*Database, error) {
+	var m T
+	name := schema.ModelName(m)
+
+	err := d.CreateDatabase(ctx, name)
+	if err != nil {
+		if !strings.Contains(err.Error(), "already exist") {
+			return nil, err
+		}
+	}
+
+	db := newDatabase(name, d)
+
+	err = db.createCollectionsFromDatabaseModel(ctx, m)
+	if err != nil {
+		return nil, err
+	}
+
+	return db, nil
+}
+
+// OpenDatabase initializes Database from Database model
+// It creates Database if necessary.
+// Creates and migrates schemas of the collections which constitutes the Database
+func OpenDatabase[T schema.Model](ctx context.Context, cfg *config.Config) (*Database, error) {
```

Comment:
My only worry with this method is if it is too confusing to have both i.e. database model as well as schema model. We have a UseDatabase API and then we have Create/Drop that can be used on the database object, in addition to that now we also have OpenDatabase API which is doing some magic.

Reply:
Database model bring many benefits:
* It allows to setup relations between tables in a declarative way, which we would need as soon as we start adding secondary indexes.
* It allows for user to have control over collection naming
* It visually and logically combines collections into single entity, which we can perform operations on. (like automatically creating collections of the db, etc...)

Reply:
We discussed this before, it is not a good idea to have a Database model. Let's not bring in changes that we discussed and decided not to go with.